### PR TITLE
X86/hex0: Shrink x86 bootstrap seeds

### PR DESCRIPTION
**Pull request by stikonas**

### CODE REVIEW

**Comment by stikonas** [comment]
@oriansj @goldsteinn review please

**Comment by goldsteinn** [comment]
Im a bit confused by this, AFAICT you are just optimizing the elf layout but I don't see any bytes actually removed....

Also just in case you havent seen it: https://nathanotterness.com/2021/10/tiny_elf_modernized.html

**Comment by stikonas** [comment]
> Im a bit confused by this, AFAICT you are just optimizing the elf layout but I don't see any bytes actually removed....

Well, there are 8 bytes commented out now. I overlapped two identical parts of the ELF header. Yes it's a bit of a hack, but it shouldn't make it much harder to understand I hope.

**Comment by stikonas** [comment]
> Also just in case you havent seen it: https://nathanotterness.com/2021/10/tiny_elf_modernized.html
 Yeah, I think I have seen it before.

**Comment by stikonas** [comment]
(updated the label address comments that I forgot to update in `hex0_x86.hex0`)

File: POSIX/x86/hex0_x86.hex0 [hunk 1]
```diff
@@ -21,33 +21,36 @@
...
 00 00                           # e_shstrndx index of the section names
+# End of ELF base header
 
-## Program Header
-#:ELF_program_headers
-#:ELF_program_header__text
-01 00 00 00                     # ph_type: PT-LOAD = 1
-00 00 00 00                     # ph_offset
+# 01 00 00 00                   # ph_type: PT-LOAD = 1
+# 00 00 00 00                   # ph_offset
```
Comment (goldsteinn) [inline POSIX/x86/hex0_x86.hex0:47]: Oh I see. Okay.

**Comment by goldsteinn** [comment]
This seems sensible to me. Did you test this?

**Comment by stikonas** [comment]
Yes, I tested it on modern Linux and on builder-hex0. Both worked.

**Comment by goldsteinn** [comment]
LGTM. I didn't retest if you'd like that lmk

**Comment by stikonas** [comment]
I wonder if it's useful to harden segment permissions though (p_flags). In hex0 we don't have any data, so we could set to to execute only. Of course in kaem we do need rwx to read and write to various variables at the end.

**Comment by goldsteinn** [comment]
> I wonder if it's useful to harden segment permissions though (p_flags). In hex0 we don't have any data, so we could set to to execute only. Of course in kaem we do need rwx to read and write to various variables at the end.

It doesn't cost any codesize, so don't see any reason not too, not that I really see any reason too either...

**Comment by stikonas** [comment]
> Also just in case you havent seen it: https://nathanotterness.com/2021/10/tiny_elf_modernized.html

Oh after reading this again, I think I can  sometimes use `mov al,` to make it smaller

**Comment by stikonas** [comment]
`mov al,` instruction instead of `push imm8; pop eax` saved another byte. Potentially we could use it in one more place, but that would break hex0 if somebody supplies more than 128 arguments.

File: POSIX/x86/hex0_x86.hex0 [hunk 2]
```diff
@@ -146,8 +149,7 @@ BE 00 00 00                     # ph_memsz
...
 	8929                        ; mov_[ecx],ebp               # Move edge to buffer
 	89FB                        ; mov_ebx,edi                 # Move outfile to ebx
-	6A04                        ; push !4
-	58                          ; pop_eax                     # Set write syscall in eax
+	B004                        ; mov_al, !4                  # Set write syscall in eax
 	CD80                        ; int !0x80                   # Do the write
-	EB B3                       ; jmp !loop_reset_all         # Start a fresh byte
-#:ELF_end ; (0x80480be)
+	EB B4                       ; jmp !loop_reset_all         # Start a fresh byte
+#:ELF_end ; (0x80480B6)
```
Comment (goldsteinn) [inline POSIX/x86/hex0_x86.hex0]: b5 now no?

**Comment by goldsteinn** [comment]
Can you make the `mov al` change a seperate commit to simplify review.

**Comment by stikonas** [comment]
@goldsteinn what is the reason for:
```
# Start of hex str to int
sub    al, 0x30
sub    al, 0x0A
```

surely we can just do `sub al, 0x3A` here with an appropriate comment. So that's another 2 bytes.

**Comment by goldsteinn** [comment]
> @goldsteinn what is the reason for:
> 
> ```
> # Start of hex str to int
> sub    al, 0x30
> sub    al, 0x0A
> ```
> 
> surely we can just do `sub al, 0x3A` here with an appropriate comment. So that's another 2 bytes.

We use ` sub al, 0x0a` to set flags. If we do `sub al 0x3a` we will treat bytes in [0, 9] as ascii values (and since they won't nicely fit in 4 bits it will cause proper bugs).

**Comment by stikonas** [comment]
> We use ` sub al, 0x0a` to set flags. If we do `sub al 0x3a` we will treat bytes in [0, 9] as ascii values (and since they won't nicely fit in 4 bits it will cause proper bugs).

Oh indeed, we do have `jb` after those...

**Comment by stikonas** [comment]
I've split x86/hex0 changes into separate commits.

I've also added a new `amd64` bootstrap seeds that follows the same algorithm. It is generally very close to `x86` version though uses slightly different registers (because of `x86_64` calling convention)

**Comment by stikonas** [comment]
amd64 version does not have an overlapping elf and program headers since the order of the fields in program header is a bit different, `p_flags` come earlier, and that field can't be non-zero. In principle we can have non-zero number in `e_shnum` too, the binaries would run (probably because `e_shentsize` is 0, so the kernel ignores `e_shnum`  field) but that is probably an implementation detail and not elf spec.  So unlike on `x86` I've decided not to overlap elf and program headers.

**Comment by stikonas** [comment]
Update to stage0-posix-amd64 is here https://github.com/stikonas/stage0-posix-amd64/tree/hex0-seed and will be merged after this review is done here.

**Pull request merged**